### PR TITLE
ISPN-8516 OSGI integration tests random failure listening to RMI port

### PR DESCRIPTION
**Pull request by danberindei**
https://issues.jboss.org/browse/ISPN-8516

Kill per-suite Karaf container in a shutdown hook.

Please cherry-pick to 9.1.x as well.

### CODE REVIEW

**Comment by galderz** [comment]
Build failed, checkstyle

**Comment by danberindei** [comment]
Removed unused import and rebased.

**Comment by danberindei** [comment]
Test failures are unrelated.

**Comment by pruivo** [comment]
@danberindei is the REST failures related?
@galderz is good to go?

**Comment by danberindei** [comment]
@pruivo Yes, this is good to go -- the REST failures were unrelated.

**Comment by danberindei** [comment]
Build #11 seems to be stuck in the `lock` module, specifically `ClusteredLockSplitBrainTest`, but I don't think it's related:

    "testng-ClusteredLockSplitBrainTest[DIST_SYNC]" #13 prio=5 os_prio=0 tid=0x00007f6c386883a0 nid=0x3ed5 waiting on condition [0x00007f6c2459c000]
    java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000000fe839a30> (a java.util.concurrent.CompletableFuture$Signaller)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.CompletableFuture$Signaller.block(CompletableFuture.java:1693)
	at java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3323)
	at java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1729)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1895)
	at org.infinispan.functional.FunctionalTestUtils.await(FunctionalTestUtils.java:42)
	at org.infinispan.lock.ClusteredLockSplitBrainTest.assertFailureFromMinorityPartition(ClusteredLockSplitBrainTest.java:169)
	at org.infinispan.lock.ClusteredLockSplitBrainTest.testAutoReleaseIfLockIsAcquiredFromAMinorityPartition(ClusteredLockSplitBrainTest.java:164)

**Comment by ryanemerson** [comment]
I aborted #11, hopefully the next build will have more luck

**Comment by tristantarrant** [comment]
I got this running locally: https://gist.github.com/tristantarrant/5df100d04bcdea4d3d84347ce1ce76ca

**Comment by tristantarrant** [comment]
The exception I saw is actually the same as the one reported in ISPN-8516

**Comment by danberindei** [comment]
@tristantarrant please try it again with `-Dverbose.karaf=true`, and send me the karaf logs if you get the failure again. I speculated that the exception you got was caused by the container leak, but I only managed to reproduce the leak on my machine, not the test failure.

**Comment by danberindei** [comment]
@tristantarrant oops, I had actually forgotten to call `field.setAccessible(true)` and so my shutdown hook wasn't doing anything. I fixed that and also made some changes to only install the hook when a `PerSuite` test is actually running (JUnit instantiates the runners before checking for categories).

File: integrationtests/osgi/src/test/java/org/infinispan/it/osgi/util/IspnKarafOptions.java [hunk 1]
```diff
@@ -333,7 +316,7 @@ public static Option commonOptions() throws Exception {
       return composite(karafContainer(),
                        vmOptions("-Djava.net.preferIPv4Stack=true", "-Djgroups.bind_addr=127.0.0.1"),
                        vmOptions("-Xmx500m", "-XX:+HeapDumpOnOutOfMemoryError", "-XX:HeapDumpPath=" + System.getProperty("user.dir")),
-//                       vmOptions("-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=8000"),
+                       vmOptions("-agentlib:jdwp=transport=dt_socket,server=n,suspend=y,address=5006"),
```
Comment (tristantarrant) [inline integrationtests/osgi/src/test/java/org/infinispan/it/osgi/util/IspnKarafOptions.java]: I think you want to comment this out again :)

**Comment by danberindei** [comment]
Sorry @tristantarrant, updated!

**Comment by tristantarrant** [comment]
Merged, thanks

**Pull request merged**